### PR TITLE
[FEATURE] Add SQL query data asset for new experimental datasources

### PR DESCRIPTION
**Pull request by billdirks**
- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/execution_engine/sqlalchemy_execution_engine.py
Comment (billdirks) [inline great_expectations/execution_engine/sqlalchemy_execution_engine.py:1154]: I am going to file a ticket to add sampling support to new experimental datasources. We should be able to use the same strategy/helper method I use here to add query support.

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (billdirks) [inline great_expectations/experimental/datasources/sql_datasource.py]: I use the `SQL_DATETIME_COUNTER` to make sure every subquery alias is unique. I could use a generated uuid in the function itself. Let me know if you prefer this or a different strategy.

File: great_expectations/experimental/datasources/sqlite_datasource.py
Comment (billdirks) [inline great_expectations/experimental/datasources/sqlite_datasource.py]: In addition to this private helper method, I tried was defining a mixin class that only defined the `add_year_and_month_splitter` that could be mixed in with the `SqliteTableAsset` and `SqliteQueryAsset`. This let me sidestep defining `add_year_and_month_splitter` on these classes. I ran into issues the with type of `self` unless I made the mixin class a subclass of `SQLAsset` but it felt strange mixing that into `SqliteQueryAsset` which is derived from `QueryAsset`.  It also made the return type of this function a generic `SQLAsset` since the mixin wouldn't know about `[Table|Query]Asset`.

The main issue is the inheritance tree looks like:

```
SQLAsset -> TableAsset -> SqliteTableAsset
                 |
                 |-> QueryAsset -> SqliteQueryAsset
```

So the `Sqlite*Asset`s earliest common ancestor is `SQLAsset`. The mixin would let me spread the functionality across the hierarchy without worrying about the structure of this inheritance tree with the drawback of losing some type specificity.

I could play with that again if that is preferred. This way is a bit more verbose but is also more specific so I decided to go with that for now.

Comment (billdirks) [reply great_expectations/experimental/datasources/sqlite_datasource.py]: It just occurred to me I could probably dynamically generate the `Sqlite*Asset` subclasses to reduce the boilerplate.
I don't want sqlite to block this PR though since the main value comes from the `SQLDatasource` changes so unless someone feels strongly or there are a lot of other changes required I may save this more a followup PR to prep for adding more splitters.

File: great_expectations/execution_engine/sqlalchemy_execution_engine.py
Comment (alexsherstinsky) [inline great_expectations/execution_engine/sqlalchemy_execution_engine.py]: @billdirks Would it make sense to utilize some kind of an SQL query parser into tokens -- or something equivalent on our part -- to allow for a flexible number `\s`, `\t`, `\r`, `\n`, etc. characters in the query string?  Thanks.

File: great_expectations/execution_engine/sqlalchemy_execution_engine.py
Comment (alexsherstinsky) [inline great_expectations/execution_engine/sqlalchemy_execution_engine.py]: @billdirks While not part of this effort, it would be great to investigate the feasibility of using the formal table columns in the `sa.select("*")` statement (analogous to the way it is done as part of introspection functions in "great_expectations/expectations/metrics/util.py") -- the advantage is that this will provide objects for columns, which could then be utilized for other operations (e.g., adding labels, etc.), which are more difficult (if not outright impossible) when using the string `"*"`.  Thanks.

File: great_expectations/execution_engine/sqlalchemy_execution_engine.py
Comment (alexsherstinsky) [inline great_expectations/execution_engine/sqlalchemy_execution_engine.py:1150]: @billdirks -- ❤️ For a simple and elegant way of bringing back the `SQL` query as a legitimate way to specify a selectable!

File: great_expectations/execution_engine/sqlalchemy_execution_engine.py
Comment (alexsherstinsky) [inline great_expectations/execution_engine/sqlalchemy_execution_engine.py]: @billdirks While equivalent, I think that the following style, used in other parts of the codebase, is more elegant in the sense that it always uses booleans to compute the mutual exclusion as a boolean (i.e., without the integers):

```
        if (
            sum(
                bool(x)
                for x in [
                    batch is not None,
                    batch_list is not None,
                    batch_request is not None,
                    batch_request_list is not None,
                ]
            )
            > 1 
        ):
            raise ValueError(
                "No more than one of batch, batch_list, batch_request, or batch_request_list can be specified"
            )
```

Your call.  Thanks for considering.

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/sql_datasource.py]: @billdirks In my opinion, something that ends on `_{str(uuid.uuid4())[:8]}` might be slightly more elegant.  We have a function in `great_expectations/util.py` called `generate_temporary_table_name()`; one can create a better-named one here by calling `generate_temporary_table_name()` with `default_table_name_prefix=""` and leaving the `num_digits` same.  Thanks.

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (alexsherstinsky) [inline great_expectations/experimental/datasources/sql_datasource.py]: @billdirks Should we not be building a formal `SQLAlchemy` query object for the given DBMS backend, and then utilizing something like the following:

```
            query: str = str(
                selectable.compile(
                    dialect=execution_engine.dialect,
                    compile_kwargs={"literal_binds": True},
                )
            )
```

in order to obtain the string-valued query version for printing?  Thanks.

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (alexsherstinsky) [inline great_expectations/experimental/datasources/sql_datasource.py]: @billdirks I think that this should be flexible in terms of blank spaces, tabs, etc. (please see an earlier comment).  Thanks.

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (alexsherstinsky) [inline great_expectations/experimental/datasources/sql_datasource.py]: @billdirks -- I ❤️ the idea of the `SQLDataAsset` defining an interface for query, table, etc. to implement their specifics for!  Thanks!

File: tests/experimental/datasources/test_integration_datasource.py
Comment (alexsherstinsky) [inline tests/experimental/datasources/test_integration_datasource.py]: @billdirks In #6981 `sql_data` is moved to a util module (the merge should not be too difficult).

File: tests/experimental/datasources/test_integration_datasource.py
Comment (alexsherstinsky) [inline tests/experimental/datasources/test_integration_datasource.py]: @billdirks In #6981 `multibatch_sql_data` is moved to a util module (the merge should not be too difficult).

File: tests/experimental/datasources/test_integration_datasource.py
Comment (alexsherstinsky) [inline tests/experimental/datasources/test_integration_datasource.py]: @billdirks Per earlier comments, I would want to be able to include a free amount of blank spaces in the `SQL` query specification.  Thanks.

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (alexsherstinsky) [inline great_expectations/experimental/datasources/sql_datasource.py]: @billdirks While not part of this change, the calculated variable names `year` and `month` are misleading.  They should be `years` and `months` based on the fact that they represent ranges.  Thank you.

File: great_expectations/experimental/datasources/sqlite_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/sqlite_datasource.py]: @billdirks I would be interested in better understanding this explanation, even if the implementation is not changed.  One thought that came to me was that in order to avoid the same code that seemingly appears in the parent `SQLAsset` class for `add_year_and_month_splitter()` would be to call `super()` here and instantiate the output field by field with the present class (similarly to the way it is done in `DataAssistant` implementations for returning the corresponding `DataAssistantResult` type (something like this technique I thought. might help here).  Or could something as simple as `cast()` work here (it seems that the output is always the same and it is just a matter of returning the object with the correct type)?  Thanks.

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@billdirks This is great stuff; I am requesting only that my recommendations are considered and questions answered.  In other words, no particular changes are required, unless you agree with the suggestions and/or any JIRA tickets for future enhancements are created.  And, of course, all required tests in Azure must pass.  Then I will be happy to approve.  Thank you!

File: great_expectations/execution_engine/sqlalchemy_execution_engine.py
Comment (billdirks) [reply great_expectations/execution_engine/sqlalchemy_execution_engine.py]: I can look more into if it's possible to do this. I would like to know also. One possible difficulty is we are selecting from an arbitrary query and there is no sql table or sqlalchemy model defined. We will get a dictionary returned with the key being the column name currently.
In past experience with django, I often found the dict preferable because, object creation there was orders of magnitude slower and with a biggish data set this really mattered. Don't know the situation here though.

File: great_expectations/execution_engine/sqlalchemy_execution_engine.py
Comment (billdirks) [reply great_expectations/execution_engine/sqlalchemy_execution_engine.py]: Added an lstrip to remove any initial whitespace characters.

File: great_expectations/execution_engine/sqlalchemy_execution_engine.py
Comment (billdirks) [reply great_expectations/execution_engine/sqlalchemy_execution_engine.py]: Done.
I didn't use `bool` since I prefer to sum `int`s but otherwise similar.

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (billdirks) [reply great_expectations/experimental/datasources/sql_datasource.py]: I can update this is a followup PR if you don't mind since this isn't part of this change and will involve some plumbing throughout this file.

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (billdirks) [reply great_expectations/experimental/datasources/sql_datasource.py]: Done

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (billdirks) [reply great_expectations/experimental/datasources/sql_datasource.py]: Done.

File: tests/experimental/datasources/test_integration_datasource.py
Comment (billdirks) [reply tests/experimental/datasources/test_integration_datasource.py]: Added prepended spaces.

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (billdirks) [reply great_expectations/experimental/datasources/sql_datasource.py]: Can you provide a code snippet and/or point me to docs about how to create a formal `SQLALchemy` object with a subselect in the from clause?

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (Kilo59) [inline great_expectations/experimental/datasources/sql_datasource.py]: 1. Is the class meant to be instantiated?
2. Will a user ever have a config file with `type: sqlasset`?

If the answer is no, then it probably doesn't need to be registered by the `MetaDatasource`.
You could make this `_SQLAsset(DataAsset)`: and this registration step would be skipped.

https://github.com/great-expectations/great_expectations/blob/1e91b803778309a4d18c27fadb156148b7109760/great_expectations/experimental/datasources/sources.py#L127-L132

However, it might be better to be more explicit about this and have some `ClassVar` `_SKIP_ASSET_REGISTRATION` attribute that gets checked. But that doesn't need to be done in this PR.

File: great_expectations/execution_engine/sqlalchemy_execution_engine.py
Comment (alexsherstinsky) [reply great_expectations/execution_engine/sqlalchemy_execution_engine.py]: @billdirks All I can think about is if the example(s) from `great_expectations/expectations/metrics/util.py:column_reflection_fallback()` and/or from `great_expectations/expectations/metrics/multicolumn_map_metrics/compound_columns_unique.py` and/or `great_expectations/execution_engine/sqlalchemy_execution_engine.py::ExecutionEngine.get_domain_records()` apply, then there may be a way; otherwise, we keep the `sa.select("*")` -- actually it looks like we resort to `sa.select([sa.text("*")])` as the catch-all in the latter.  So, perhaps worth a quick check, just in case.  Thanks!

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/sql_datasource.py]: @billdirks Examples that come to mind are:
* In `great_expectations/expectations/metrics/map_metric_provider.py`: `_sqlalchemy_map_condition_unexpected_count_value()` and `_sqlalchemy_map_condition_unexpected_count_aggregate_fn()` (among others).
* In `great_expectations/expectations/metrics/column_aggregate_metrics/column_quantile_values.py`, methods called by `_sqlalchemy()`.

Happy to discuss.  Thank you!

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/sql_datasource.py]: 🙇

File: tests/experimental/datasources/test_integration_datasource.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/test_integration_datasource.py]: 🙇

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (billdirks) [reply great_expectations/experimental/datasources/sql_datasource.py]: Good call out. I had sketched this before I saw you exclude the `_` prefix assets and have added it.

File: great_expectations/experimental/datasources/sqlite_datasource.py
Comment (billdirks) [reply great_expectations/experimental/datasources/sqlite_datasource.py]: Yes the issue is with typing. I want to add a common function to the 2 sqlite data assets but they don't share a direct superclass. I also want mypy to throw an error when this function is used incorrectly. I really don't like the duplication though. Instead of a `cast`, I've updated this to use a mixin where I can make an assert about types so we at least get a runtime check with a good error message.

File: great_expectations/execution_engine/sqlalchemy_execution_engine.py
Comment (billdirks) [reply great_expectations/execution_engine/sqlalchemy_execution_engine.py]: Those examples don't work here and they fall back to `select(*)` for similar use cases.
I think it would be hard to do because we would need to parse the subselect query, which could have a `*` in it, and then map the selected columns back to their origin to infer types.

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (billdirks) [reply great_expectations/experimental/datasources/sql_datasource.py]: Thank you! This is done.

File: great_expectations/execution_engine/sqlalchemy_execution_engine.py
Comment (alexsherstinsky) [inline great_expectations/execution_engine/sqlalchemy_execution_engine.py:1232]: Nice "XOR"! :)

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (alexsherstinsky) [inline great_expectations/experimental/datasources/sql_datasource.py:151]: @billdirks This is great now how it is all `SQLAlchemy` object form.  One last thing on this, please: I believe that it is safer to explicitly convert `query` into `str`, for example:
```
query_string = query.compile(compile_kwargs={"literal_binds": True})
```

Thank you.

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@billdirks Thank you for this -- really good stuff.  Here are a couple of remaining things:
* Static Type Checks are currently failing (probably easy to address);
* I made a suggestion to explicitly compile the query into a string (I feel that this is important);
* Optional: make one of the SQL query tests include some blank spaces/tabs around "SELECT" (sorry if you already added one, and I missed it in the code). 😄 

Thank you!

**Review by Kilo59 — APPROVED** [review]

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (billdirks) [reply great_expectations/experimental/datasources/sql_datasource.py:151]: Done. Could you explain why you think calling compile directly is better?

**Review by billdirks — COMMENTED** [review]
@alexsherstinsky Thanks for your comments. This is ready for re-review.

* static checking now succeeds. It was passing locally so must be a mypy version issue or something related.
* I now compile the sql query. I'm wondering what makes this safer? I googled around but couldn't find any conversations about this.
* There is one: https://github.com/great-expectations/great_expectations/pull/6999/files#diff-b97e96dba34432dc6b4f3bb19ce37d8e23693cbbcf5a6e244f4803bc494fa2c9R100

File: great_expectations/experimental/datasources/sql_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/sql_datasource.py:151]: As per earlier versions of `SQLAlchemy`, it has been utilized as a more explicit way to control what happens in "to string" operation.  It is certainly worth seeing how this approach has changed as the newer `SQLAlchemy` versions have been including more and more improvements.  Thanks.

**Review by alexsherstinsky — APPROVED** [review]
Approving provisionally -- for when all the test failures have been resolved.  Thanks!

**Pull request merged**
